### DISCUSSION TITLE
Multiple roles for SaaS

### DISCUSSION THREAD

**Discussion by danieljwestman** [directus/directus#9682] [Feature Requests]
We are trying out some alternatives in Directus for a pretty big SaaS.

Two of the main options:

1. One role for each tenant (tenant id on the role) or
2. Tenant id directly on the user object

Then use the new dynamic user/role variables to permit the correct content :

1. $CURRENT_ROLE.tenant_id or
2. $CURRENT_USER.tenant_id

The option we are leaning towards is number 1. The drawback is that it will be a lot of roles, could be hundreds or even thousands... It shouldn't be a performance issue (right?) but the current UI is not suited for this all the way. The roles list under Settings and the roles navigation under Users is example of this.

Any ideas and thoughts? 

@benhaynes @rijkvanzanten do you have any plans or mockups?

Thanks!

**Comment by benhaynes** [comment]
There's no functional issue with many many roles, but we might want to update the UI to more efficiently handle a large number of roles. No mockups for this, but I imagine it will include using the `dense` list type and a search bar, much like we do for the collections navigation.

**Reply by danieljwestman** [reply]
> There's no functional issue with many many roles, but we might want to update the UI to more efficiently handle a large number of roles. No mockups for this, but I imagine it will include using the `dense` list type and a search bar, much like we do for the collections navigation.

Yes 👍🏼 

So there are two places this must be updated; in the Roles list under Settings module and in the side navigation under Users module. Missing something? Maybe I can sponsor this. Let's talk?

**Reply by benhaynes** [reply]
Sure thing, we can coordinate over Discord. 👍

**Reply by azadzain** [reply]
What about a roles grouping like in the data model groups?

**Comment by w0kyj** [comment]
Also note that use case include:
- Support for between 30 to 500 users per role

This means the Settings-->Role-->Users in Role would need to be updated also to handle large numbers of users assigned to a role -- pagination and search would be useful.

**Reply by rijkvanzanten** [reply]
Users currently uses regular layouts, so that's already covered 👍🏻

**Comment by xu4wang** [comment]
I am prototyping a similar SaaS product. However, in my scenario, each tenant also has several different roles. 

For example, we have 1000 merchants( tenants), each merchants may have several  log in accounts(Alice, Bob, Tom) belongs to several different roles(cashier, manager).   Alice and Bob are cashiers, Tom is the manager.  They have role based different access rights.

**Comment by xu4wang** [comment]
Hello @danieljwestman 

May I know why you choose option 1 instead of option 2 ?
I just did some test using option 2 and was satisfied with it. :-)

The changes I made
1. added a Tenants collection, 
2. added one field tenant_id in the user table, the value is the item.id in Tenants
3. for every collections need multi tenant feature, add a M2O field tenant_id, 
4 add role based rules for read/write/update/delete records with tenant_id field.  
4.1 Add preset for creating item, using $CURRENT_USER.tenant_id as the value of newly created record.
4.2 Add rule for reading, can only access the records when tenant_id equals to  $CURRENT_USER.tenant_id.
...

So far so good, everything looks fine.

What's the drawback of this option 2?

Best Regards,Austin

**Reply by piotr-cz** [reply]
This worked great for my setup, thank you!
Except the point 4.1. After I've added **Field Preset** to related collection (*Manager -> News -> Create*)

```json
{
    "organisation_id": "$CURRENT_USER.organisation_id"
}
```

the **Create Item** forms shows a `[Forbidden] You don't have permission to access this.` message in popup.
The API request is

```http
GET 403 /items/organisations/%24CURRENT_USER.organisation_id?fields[]=name&fields[]=id

Content-Type: application/json; charset=utf-8
{"errors":[{"message":"You don't have permission to access this.","extensions":{"code":"FORBIDDEN"}}]}
```

I have no idea what I'm doing wrong.
The `$CURRENT_USER.organisation_id` dynamic variable isn't substituted in API request (actually any dynamic variable isn't), however executing the query that is output with `LOG_LEVEL="trace"` works fine:

```sql
SELECT
  `directus_users`.`organisation_id`,
  `directus_users`.`id`
FROM `directus_users`
WHERE `directus_users`.`id` = 'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx'
ORDER BY `directus_users`.`id` asc LIMIT 100
```

According to Docs > Filter Rules > [Dynamic Variables](https://docs.directus.io/configuration/filter-rules/#dynamic-variables) this should work:

> When configuring permissions, `$CURRENT_USER` and `$CURRENT_ROLE` allow you to specify any (nested) field under the current user/role as well as the root ID. For example: `$CURRENT_ROLE.name` or `$CURRENT_USER.avatar.filesize`. This includes custom fields that were added to the directus_users/directus_roles tables.
> Note: This feature is only available for permissions, validation, and presets. Regular filters and conditional fields currently only support the root ID.



Same dynamic variable works fine for *Item Permissions*

I've tried adding Read permissions to Organisation by checking in *Role > Directus Users > Read > Field Permissions > Organisation* but this doesn't change anything.

**Reply by piotr-cz** [reply]
I've traced the problem down to https://github.com/directus/directus/blob/cd58611aa7114dd7ef13f0819f77f9858f34ad1e/api/src/services/items.ts#L311-L316

where the `key` variable is assigned to `$CURRENT_USER.organisation_id` and throws inside `validateKeys` function (which compares it's value to `number`).

Somehow somewhere the **Dynamic Variable** isn't replaced by real value.

___
Update:
Apparently this feature has been removed without update in docs in https://github.com/directus/directus/pull/10576#issuecomment-1003481572

**Reply by pkro** [reply]
>     1. added a Tenants collection,
> 
>     2. added one field tenant_id in the user table, the value is the item.id in Tenants
> 
>     3. for every collections need multi tenant feature, add a M2O field tenant_id,
>        4 add role based rules for read/write/update/delete records with tenant_id field.
>        4.1 Add preset for creating item, using $CURRENT_USER.tenant_id as the value of newly created record.
>        4.2 Add rule for reading, can only access the records when tenant_id equals to  $CURRENT_USER.tenant_id.

This is also my setup minus the preset, which I now added (thanks! simply didn't find where and was stuck trying to use default value, which doesn't support dynamic variables).

My problem now is that the user can still overrule the preset, potentially creating an item in another tenant. I can use a filter to force the tenant to only be able to create entries with their own tenant_id, but then the request needs to contain the tenant ID, making the preset "useless" as i can't remove the client field from all my requests like I intended.

Any way around this?

**Reply by piotr-cz** [reply]
@pkro 
As for the *Field Preset*, I filled issue #14315

To restrict user access to (s)his tenant/ organization (also in related M2O collections), set up the Role > Permissions > Tenant as follows:
- Create, Delete, Share: *No access*
- Read: *Use Custom*:
  - Item Permissions > *ID Equals `$CURRENT_USER.tenant_id`*
  - Field Permissions > fields you'd like user to see (probably all)
- Update: *Use Custom*:
  - Item Permissions > *ID Equals `$CURRENT_USER.tenant_id`*
  - Field Permissions > fields you'd like user to be able to edit

I think a recipe in Directus docs with all the steps would be really helpful.
Setting this all up requires quite a few steps.

PS: I haven't tried setting up a scenario where user may belong to multiple tenants.

**Reply by rijkvanzanten** [reply]
(Cc @erondpowell)

**Reply by pkro** [reply]
Edit: solved, I can of course remove write access for the client field in all collections and set it automatically using the preset, so everything should work as intended and no validation will be necessary.

Thanks @piotr-cz for taking the time to answer this, much appreciated! 

(client = tenant in the following):
 
That's about the setup that I have now, and this works and avoids all permission issues. I have client as a m2o relationship in all my collections.

What I wanted to get rid of now was the necessity having to pass the client ID in my API requests at all, because I could set it with the preset (this works when using the API it seems). But when I use only the preset, a malicious request could override the preset, setting the client ID to the ID of another client. When I try to restrict that by using "client equals $CURRENT_USER.client.id" in the write/create permissions of a collection, the request fails because the client field is not present in the API request. So it seems I can either have an auto-filled client ID for API requests OR a check that the client ID is set correctly, but not both.

In short: "Field validation" doesn't take "field presets" into account.

**Comment by emilyf** [comment]
Hello @xu4wang @danieljwestman ! I'm trying to decide which of these routes to go and wondering if you have any particular issues that have come up related to approach 1 vs. approach 2. 

It seems option 2 is easier to implement and less overhead? My situation is similar to @xu4wang scenario of

> For example, we have 1000 merchants( tenants), each merchants may have several log in accounts(Alice, Bob, Tom) belongs to several different roles(cashier, manager). Alice and Bob are cashiers, Tom is the manager. They have role based different access rights.

**Reply by xu4wang** [reply]
i vote for option 2 also.

But if your app is with lots of features, you may consider the multi project solution too.  Multi projects solution in my opinion is most flexible but need invest on projects management.

**Comment by u12206050** [comment]
I have been using option 2 for quite a while, although have now been met by a case I am unable to currently support, which is the user having access to multiple tenants, with / without different roles.

We are running a non profit event system, and club managers can only view their own events and their own members. This works fine with option 2. However two of the club managers are also on our support team which should have access to all clubs and members within a specific region. Also, there are a few additional things support can do, that a normal club manager can not.

This lead me to think of setting up my own roles collection, with a module to allow a user to switch one of their other roles, which causes an update on their user profile and then refreshes the page, loading in with the newly selected role and optional other settings (ie. other clubs)

Here is an idea of dropdown that could show up if it is detected the current user has multiple roles.
![Screenshot 2022-03-30 at 08 35 36](https://user-images.githubusercontent.com/6641242/160766798-d0b9858c-ae08-4e8e-9b1a-489b3fb20106.png)

**Reply by xu4wang** [reply]
@u12206050 Thanks for sharing the idea of switching roles.

In my use case, there are actually two dimensions  for access control.
1. Role, which is defined by directus role the user has
2. Tenant, which is a ID I added to the system collection directus user.

If there is a people need to play same role in different tenant(club in your case), then there should be some logic for the user to select which tenant he want to access, and then change the value of tenant id inside directus user table.

For more complicated scenario, if the user plays multiple roles in multiple tenants,  then may need to change both role id and tenant id in his/her profile.

**Reply by tspvivek** [reply]
@xu4wang  Updating tenantid in user table will create confusion,  if a user is using two different browsers/devices simultaneously.

**Reply by tspvivek** [reply]
We are using directus as API server and built a SAAS app with complex scenario like user with multiple roles in multiple organisations and multiple teams. We were changing core directus authenticate method for this but thanks to directus team, now we are able to use the new authenticate hook to achieve this.

First we need to create a  membership table with 4 fields (user, organisation, team and role having M2O relation to respective collections)
Then in each request sent to directus, add membershipId in params (which must be the id of the membership table).  You can also send in header according to your convenience.

In authenticate hook check if the membershipId is present for the user and update accountability with organisation, team and role from the  membership table like in below code.

We have added two dynamic variables in directus core, CURRENT_ORGANISATION (resolves to accountability.organisation) and CURRENT_TEAM (resolves to accountability.team) to use it in permissions.

Note: All registered users (except administrators) should have authenticated role in directus_users table.

```
const membership = await checkTeam(membershipId, database, user.id);
accountability.organisation = (membership && membership.organisation) || null;
accountability.team = (membership && membership.team) || null;
accountability.role = (membership && membership.role) || user.role;

accountability.user = user.id;
accountability.admin = user.admin_access === true || user.admin_access == 1;
accountability.app = user.app_access === true || user.app_access == 1;
```

Hope this helps someone !!

**Reply by u12206050** [reply]
@tspvivek This sounds very good, have a few questions if you don't mind.

1. How did you add the two addition dynamic variables, and do they work using them in permissions?
2. Were you able to also make use of this from the Directus app, or only your own custom frontend?

**Reply by tspvivek** [reply]
1. I have added dynamic variables by modifying these two files.
@directus/shared/dist/cjs/utils/is-dynamic-variable.js
@directus/shared/dist/cjs/utils/parse-filter.js
Sample [files](https://github.com/directus/directus/files/8503827/Dynamic.variables.zip) attached.
This is a temporary workaround till $CURRENT_ACCOUNTABILITY is available.

2. I am using NextJS app as frontend with [Refine & Directus dataprovider](https://github.com/tspvivek/refine-directus), For directus app, we should wait for directus team.

**Reply by moseskamau338** [reply]
Hi @tspvivek, did you clone Directus repo? Or how did you access those files coz I can't seem to find them in the dockerized version. How have you setup your deployment so conveniently?


Thanks for the info!

**Reply by tspvivek** [reply]
I am using self hosted version, not dockerized, [like this](https://docs.directus.io/self-hosted/cli.html#server). Also i have the modified files in folder "customauth" and copying the files to node_modules during startup, my scripts are like this.

```
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1",
    "local": "cp -r customauth/* node_modules/ && CONFIG_PATH=\"config.js\" NODE_OPTIONS=\"--trace-warnings\" npx directus start",
    "staging": "cp -r customauth/* node_modules/ && CONFIG_PATH=\"config.stg.js\" npx directus start",
    "development": "cp -r customauth/* node_modules/ && CONFIG_PATH=\"config.dev.js\" npx directus start",
    "production": "cp -r customauth/* node_modules/ && CONFIG_PATH=\"config.prod.js\" npx directus start"
  }
```

**Reply by moseskamau338** [reply]
Thanks a lot!

This is so helpful. I will give it a try soon.

**Reply by moseskamau338** [reply]
Hi @tspvivek, I have implemented your approach, however, the dynamic variables are not being injected into the system. Here is what I have:
*Directus Version* - `10.5.3`

Custom Auth Hook:

```ts
export default defineHook(({ filter }, {database}) => {
	filter("authenticate", async (accountability: any, { req }) => {

		async function getMembership(id: string){/*... logic for fetching membership details ....*/}

		  const rawToken =
				req?.headers.authorization ||
				req?.headers?.["x-access-token"] ||
				req?.body?.token ||
				req.query?.token;

		console.log('Raw Token: ', rawToken);

		const token = rawToken ? rawToken.split(" ")[1] : false;
		const decodedToken = token ? jwt.decode(token) : false;
		console.log('Decoded token: ', decodedToken);

		if (token && decodedToken) {
			// continue with existing user

			// if is admin and app access, pass
				// else: update accountability
			if (!decodedToken.admin_access && decodedToken.app_access && decodedToken.role !== '<role-id>') 
			{
				//user is ! admin	
				if (req?.headers?.membership_id && await getMembership(req?.headers?.membership_id)) 
				{
					let membership_entry =  await getMembership(req?.headers?.membership_id)
					//update accountability
					accountability = {
						...accountability,
						membership_entry,
						user: membership_entry.user_id,
  						role: membership_entry.user_role,
						account_id: membership_entry.account_id,
						account_type: membership_entry.account_type,
						account_slug: membership_entry.account_slug,
					}

					console.log('New accountability: ', accountability);
					
				}else{
					throw new BaseException("Membership ID required", 401, "INVALID_CREDENTIALS");
					
				}

			}

		}

	
		console.log('Accountability: ', accountability);
	
		return accountability;
	});

});
```

This works fine and I can see it(accountability object) in an example flow.

However, the dynamic variables are not being injected at all. I also noticed that in the node modules folder, there is no `@directus` directory with the files you specified earlier.

Am I missing something? or should I find a way to use the `accountability` object directly in the role/permission filters?


Thank, any help is welcome

**Reply by tspvivek** [reply]
Files are moved to` node_modules/@directus/utils/dist/shared` directory in new versions. Without changing the files parse-filter.js and is-dynamic-variable.js, $CURRENT_ACCOUNTABILITY will not work.

**Reply by moseskamau338** [reply]
I changed the folder structure correctly as shown below:

![Imgur](https://i.imgur.com/BahMloG.png)

But my variables are not present in within Directus filters and flows


```js
// is-dynamic-variable.js

"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isDynamicVariable = void 0;
const dynamicVariablePrefixes = ['$NOW', '$CURRENT_USER', '$CURRENT_ROLE', '$CURRENT_ACCOUNT', '$CURRENT_ACCOUNT_ROLE', '$CURRENT_MEMBERSHIP'];
function isDynamicVariable(value) {
    console.log('Checking current memberships: ', value);
    return typeof value === 'string' && dynamicVariablePrefixes.some((prefix) => value.startsWith(prefix));
}
exports.isDynamicVariable = isDynamicVariable;
```

Here is my other file:

```js
// parse-filter.js

function parseDynamicVariable(value, accountability, context) {
    var _a, _b, _c;
    if (value.startsWith('$NOW')) {
        if (value.includes('(') && value.includes(')')) {
            const adjustment = (_a = value.match(constants_1.REGEX_BETWEEN_PARENS)) === null || _a === void 0 ? void 0 : _a[1];
            if (!adjustment)
                return new Date();
            return (0, adjust_date_1.adjustDate)(new Date(), adjustment);
        }
        return new Date();
    }
    if (value.startsWith('$CURRENT_USER')) {
        if (value === '$CURRENT_USER')
            return (_b = accountability === null || accountability === void 0 ? void 0 : accountability.user) !== null && _b !== void 0 ? _b : null;
        return get(context, value, null);
    }
    if (value.startsWith('$CURRENT_ROLE')) {
        if (value === '$CURRENT_ROLE')
            return (_c = accountability === null || accountability === void 0 ? void 0 : accountability.role) !== null && _c !== void 0 ? _c : null;
        return get(context, value, null);
    }

    // TEAM FIX
    console.log('Assigning custom globals: ', value, accountability);
    if (value === '$CURRENT_ACCOUNT')
        return (accountability === null || accountability === void 0 ? void 0 : accountability.account_id) || null;
    if (value === '$CURRENT_ACCOUNT_ROLE')
        return (accountability === null || accountability === void 0 ? void 0 : accountability.membership_entry?.role) || null;
    if (value === '$CURRENT_MEMBERSHIP')
        return (accountability === null || accountability === void 0 ? void 0 : accountability.membership_entry) || null;
}
```

**Comment by development-vargamarcel** [comment]
What about the idea of identities? 

User can have multiple identities, he will create a new identity for every business and share it's ID with the business to give him a role (or he can have a list of businesses that can view and select said identity, and assign a role or multiple roles inside the business) . 

User will be able to select multiple identities as 'in use' an will act as that array of identities command.

Didn't think this through very much,but how does the idea sound?
I'd apreciate some expert's view on this idea and it's shortcomings.

@tspvivek I would really appreciate your help. Your idea looks very good but I am really a beginner and I don't think I can implement all those details. That is why I am looking for a simpler solution.

**Comment by tspvivek** [comment]
@development-vargamarcel Yeah, Might looks complex because we need to add dynamic variables in core, I have asked for $CURRENT_ACCOUNTABILITY to make it simpler in this discussion https://github.com/directus/directus/discussions/12534

**Comment by jbeckton** [comment]
SaaS does not always mean everyone shares the same instance as tenants. I am considering deploying an instance of Directus per Tenant. Between the app instance and the database behind it, depending on where you deploy it you can probably keep your costs low by running the Directus app on serverless ( Google Cloud Run ). This has much better security then trying to hack your way into a single multi-tenant instance. Since Directus has an API for everything you can create a centralized management app that could be used to manage all your Directus instances via their API. Most if not all of the provisioning of new instances can be automated fairly easily.

**Reply by tecnicosya** [reply]
How many instances will support 1 server in this way? I ask because I'm thinking on something similar but get small amount of tenants running on this way.

**Comment by u12206050** [comment]
Just putting my solution out there for others to use: https://github.com/u12206050/directus-extension-role-chooser

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

---

That being said, I believe most upvotes and subscribers here come partially from https://github.com/directus/directus/discussions/4959, so lets keep these discussions in mind for that as well 🙂
